policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/config/AkkaRunnableGraph.java

```diff
@@ -70,68 +73,64 @@ KafkaProducerService kafkaProducerService(
-                                .thenApply(resource -> new Pair<>(messageAndRequest.first(), Optional.of(new Pair<>(policyRequest, Optional.ofNullable(resource))))))
-                        // If is a START/END of stream then treat as accessible
-                        .orElse(CompletableFuture.completedFuture(new Pair<>(messageAndRequest.first(), Optional.empty()))))
+                .mapAsync(PARALLELISM, messageAndRequest -> service.getResourceRules(messageAndRequest.second())
+                        //need to change this only apply if we have rules and if we have a PolicyRequest
+                        .thenApply(auditable -> PolicyServiceHierarchyProxy.applyRulesToResource(auditable))
+                        .thenApply(modifiedAuditable -> Pair.create(messageAndRequest.first(), modifiedAuditable)))
 
-                // Now we get the record level rules for all resources that weren't course grain filtered out
-                .mapAsync(PARALLELISM, (Pair<CommittableMessage<String, PolicyRequest>, Optional<Pair<PolicyRequest, Optional<LeafResource>>>> messageRequestResource) -> messageRequestResource.second()
-                        // If this is a proper request (not start or end of stream)
-                        .map(requestAndResource -> requestAndResource.second()
-                                // If there were resource rules and the resource was not redacted (course grain filtering)
-                                .map(resource -> service.getRecordRules(resource)
-                                        .thenApply(rules -> Optional.of(Optional.of(PolicyResponse.Builder.create(requestAndResource.first()).withResource(resource).withRules(rules)))))
-                                // If the resource was redacted
-                                .orElse(CompletableFuture.completedFuture(Optional.of(Optional.empty()))))
-                        // If this is a start or end of stream
-                        .orElse(CompletableFuture.completedFuture(Optional.empty()))
-                        // Keep track of original committable
-                        .thenApply(response -> new Pair<>(messageRequestResource.first(), response)))
+                // Get the record level rules for all resources that weren't course grain filtered
+                .mapAsync(PARALLELISM, (Pair<CommittableMessage<String, PolicyRequest>, AuditablePolicyResourceResponse> messageAndModifiedRequest) ->
+                        service.getRecordRules(messageAndModifiedRequest.second())
+                                //check to see if the first service request threw an exception and if so deal with it
+                        .thenApply(modifiedResource -> modifiedResource.chain(messageAndModifiedRequest.second().getAuditErrorMessage()))
+                                .thenApply(response -> Pair.create(messageAndModifiedRequest.first(), response))
```

Comment:
The definition of the lambda element isn't strictly needed as there are no curly braces `{}` around the lambda code, unless sq says otherwise.
```suggestion
                .mapAsync(PARALLELISM, messageAndModifiedRequest ->
                        service.getRecordRules(messageAndModifiedRequest.second())
                                //check to see if the first service request threw an exception and if so deal with it
                                .thenApply(modifiedResource -> modifiedResource.chain(messageAndModifiedRequest.second().getAuditErrorMessage()))
                                .thenApply(response -> Pair.create(messageAndModifiedRequest.first(), response))
```